**Table of Contents:**

* One-Tailed Tests About a Population Mean :when SIGMA KNOWN

* Two-Tailed Tests About a Population Mean:When SIGMA KNOWN

* Two-Tailed Tests About a Population Proportion

* Population Means Independent samples when variance 1 & 2 are KNOWN

* Variace Unknown and assuned EQUAL

* Variace Unknown and assuned UNEQUAL

* Dependent Samples

* Inferences About the Difference B/W Two Population Proportions

* Hypothesis Tests for Two Variances

* Z test v/s T test 

In [1]:
#import necessary libraries
import scipy 
from scipy import stats

In [2]:
#Hypothesis testing can be used to determine whether a statement about the calue of a population parameter should or should not 
#rejected.

#The null hypothesis,denoted hy Ho, is a tentative assumption about a population parameter
#the alternative hypothesis,denotated by Ha, is the opposite of what is stated in the null hypothesis

It is not always obvious how the null ans alternative hypotheses should be formulated.in some cases it is easier to identify the alternative hypothesis first,in other cases the nulll is easier

**Alternative Hypothesis as a Research Hypothesis**
 -Many applications of hypothesis testing involves an attempt to gather evidence in support of a research hypothesis,in such cases ,it is often best to begain with the alterntive hypothesis,and make it hte conclusion that the researcher hopes to support
 
 **Null Hypothesis as an assumption to be challenged**
    -we might begin with a belief or assumption that a statement about the value of a population parameter is true,we than using a hypothesis test to challenge the assumption and determine if there is statistical evidence to conclude that the assumption is incorrecet
    -in these sithuations,it is helpful to develop the null hypothesis first.
                    


**Upper-Tail Test :**     H0 : mu1 - mu2 <= 0

                  H1 : mu1 - mu2 > 0
                            
**Lower-Tail Test :**     H0 : mu1 - mu2 >= 0
 
                  H1 : mu1 - mu2 < 0
                          
                          
**Two-Tail Test :**       H0 : mu1 - mu2 = 0 

                  H1 : mu1 - mu2 !=0

# One-Tailed Tests About a Population Mean :when SIGMA KNOWN


In [3]:
#Q:- the mean responce times for a random sample of 30 pizza Deliveries is 32 minutues,

#The population standard deviation is belived to be 10 minutes.

#The pizza delivery services director wants to perform a hypothesis test, with alpha=0.05 level of significance,to determine
# whether the service goal of 30 minutes or less is being achieved.

In [4]:
#formulate H0 & H1

H0 : mu <= 30  

H1 : MU > 30  

In [5]:
#from above we can sat that this is a right tailed test

In [6]:
#given 
x=32   #sample mean
mu=30
std=10
n=30
alpha=0.05

In [7]:
import numpy as np

In [8]:
z=(x-mu)/(std/np.sqrt(n))

In [9]:
z

1.0954451150103321

In [10]:
# z-calculateted value = 1.0954451150103321

# P-Value approach

In [11]:
from scipy import stats

In [12]:
#given one is right-tailed teat so use 1-stats.norm

p_value=1-stats.norm.cdf(1.0954451150103321)
p_value

0.13666083914614902

In [13]:
#p-value is 0.13666083914614902, and given alpha value is 0.05, 

#p_value > alpha_value  [0.13666083914614902 > 0.05], so don not reject the NULL Hypothesis.

# Critical Value Approach

In [14]:
#Given alpha = 0.05

#Z-Table value at 0.05 is

z_table_value=stats.norm.ppf(0.95)
z_table_value

1.6448536269514722

In [15]:
# z_calculated = 1.0954451150103321  AND  z_table_value = z_table_value


#HERE :- z_table_value >= z_calculated [ 1.6448536269514722 >= 1.0954451150103321], we do not reject H0

**CONCLUSION**

       There are not sufficient statistical evidence to infer that Pizza delivery services is not meeting the responce goal of 30 minutes

# Two-Tailed Tests About a Population Mean:     when SIGMA KNOWN

In [16]:
# Q:- Assume that a sample of 30 milk carton provides a sample mean os 505ml.
#the populaton standard deviation is believed to be 10ml.
#perform a hypothesis test, at the 0.03 level of significanve, population mean 500ml, and to help determine whether the filling 
#process should continue operating or stopped and corrected.

In [17]:
#formulate the H0,H1

**H0**  :mu = 500 [mean filling quantity is 500, i.e no need to fill now, we can stop operating ]

**H1**  :mu != 500 [mean filling quantity is not 500, i.e we need to continue opeating]

In [18]:
#Given one is a two tailed test

In [19]:
#GIVEN 
X = 505
mu = 500
sigma = 10
n = 30 
alpha = 0.03

In [20]:
Z = (X-mu)/(sigma/np.sqrt(n))
Z

2.7386127875258306

# P-value Approach for Two-Tailed Test About a Population Mean:  when SIGMA KNOWN

In [21]:
p_value_2 = (1-stats.norm.cdf(2.7386127875258306))*2    # *2 is because we are having two-tailed tests.

p_value_2

0.0061698993205441255

In [22]:
#HERE:-  p_value_2 <= alpha [ 0.0061698993205441255 < 0.03 ]  # so reject H0

# Critical Value Approach for TWO-Tailed Test About a Population Mean :when SIGMA KNOWN

In [23]:
#Firstly Determine the critical value and rejection rule, for alpha/2 = 0.03/2  [because it is a Two-Tailed Test]

In [24]:
z_value_2=stats.norm.ppf(0.03/2)

In [25]:
z_value_2

-2.1700903775845606

In [26]:
#rejection rule:

#Reject H0 if z < -2.17  OR  z > +2.17

In [27]:
# here Z_calculate > + zvalue_2 [ 2.7386127875258306  > 2.1700903775845606 ]  so  reject HO

**CONCLUSION**

      There is no sufficient statistical evidence to infer that the null hypothesis is true (i.e the mean filling quantity is not 500ml), reject H0
      
              mean filling quantity is not 500, i.e we need to continue opeating

# One-Tailed Test About a Population Mean: when SIGMA UNKNOWN

In [28]:
#NOTE:- When sigma is unknown go for T-Test

In [29]:
#Q:- In a ice cream parlor at IIT Roorkee,the following data represent the number of ice-creams sold in 20 days
#Tests Hypothesis H0 <= 10
X2=[13,8,10,10,8,9,10,11,6,8,12,11,11,12,10,12,7,10,11,8]

In [30]:
#Given one is a Right-Tailed- Test

In [31]:
from scipy import stats

In [32]:
#it takes two inputs data & mean

stats.ttest_1samp(X2,10)  # data,mean 

Ttest_1sampResult(statistic=-0.35843385854878496, pvalue=0.7239703579964252)

In [33]:
import numpy as np
np.mean(X2)

9.85

In [34]:
len(X2)

20

In [35]:
#having a dought here

In [36]:
#having a doughts here

In [37]:
#having a doughr here

In [38]:
#having a dought here

In [39]:
#having a dought here

# Two-Tailed Tests About a Population Proportion

In [40]:
#NOTE:- in Proportion we only use Z-Test statistics

In [41]:
#Q:- For a new Year's week,the city Traffic police claimed that 50% of the accidents would be caused by drunk driving.

#A sample of 120 accidents showed that 657 were caused by drunk driving.Use these data to test the Traffic police's claim 
#with alpha = 0.05

In [42]:
#formulate the Null & Alternate Hypothesis

**H0** :- p = 0.5

**H1** :- p != 0.5

# P-value Approach for Two-Tailed Test about a Populaton Proportion

In [43]:
#for proporton import proportion_ztest

In [44]:
from statsmodels.stats.proportion import proportions_ztest

In [45]:
#Given 
count = 67
samplesize=120
p=0.5

In [46]:
proportions_ztest(count,samplesize,p)

(1.286806739751111, 0.1981616572238455)

**Test Statistic**   
    
                Z =(P - p0)/root(p0(1-p0)/n)

**Conclusion** 
   
         p_value > alpha value so we do not reject the H0.
         
     95% probabilly we can say that 50% of the accidents would be caused by drunk driving

# Hypothesis Testing: Two sample test

**Population Means**  **Independent samples when variance 1 & 2 are KNOWM**

In [47]:
#Test-statisitics is Z-value

In [48]:
# Q:- A product developer is interested in reducing the drying time of a primer paint,Two formulations of the paint are tested;
#formulation1 is the standard chemistry,and formulation 2 has a new drying ingredient that should reduce the drying time.
#from experience, it is known that the standard deviation of drying time is 8 minutes, and this inherent variability should be 
#unaffected by the addition of the new ingredient.

#Ten specimens are painted with formulation 1, and another 10 specimens are painted with formulation 2;
#the 20 specimeans are painted in random order.

#The Two-Sample average drying times are X1=121 minutes and X2 = 112 minutes,respectivily.

#What conclusions can the product developer draw about the effectiveness of the new ingredient,using alpha = 0.05?

In [49]:
#formulate the H0 & H1

**H0 :-**  mu1 = mu 2 [Means are equal , i.e new ingredient do not reduce the mean drying time].

**H1 :-**  mu1 != mu2 [adding the new ingredient to the paint significantly reduce the drying time].

In [50]:
#import necassary libraries 
import numpy as np
import math 
from scipy import stats

In [51]:
#Here given
X1=121
X2=112
sigma1=8
sigma2=8
n1=10
n2=10

In [52]:
#define a function to get Z and P values

def z_and_p(x1,x2,sigma1,sigma2,n1,n2):
    z=(x1-x2)/math.sqrt((sigma1**2/n1)+(sigma2**2/n2)) #formula to get Z_calculated_value
    if(z<0):
        p=stats.norm.cdf(z)
    else:
        p=1-stats.norm.cdf(z)
        
    print(z,p)

In [53]:
z_and_p(121,112,8,8,10,10)

2.5155764746872635 0.00594189462107364


In [54]:
stats.norm.ppf(0.95)

1.6448536269514722

**Critical Value Approach  :-**

          Z_calculated_value > z_table_value, [2.5155764746872635 > 1.6448536269514722], so reject the H0,
          
          
and we can conclude that adding the new ingredient to the paint significantlt reduces the drying time




**P-Value Approach  :-**

        P_Value < alpha, [0.00594189462107364 < 0.05], so reject the H0 
        
       AND can conclude There is evidence of a difference in means.

# Variace Unknown and assuned EQUAL

**Population variance sigma1^2&sigma2^2 Unknown, Assumed equal**

In [55]:
import pandas as pd
from scipy import stats
df4=pd.read_excel('C:/Users/jagadeesh baman/Documents/alldata/sigma_unknow_Aequal.xlsx')
df4

Observation_Number  Catalyst_1  Catalyst_2
0                   1       91.50       89.19
1                   2       94.18       90.95
2                   3       92.18       90.46
3                   4       95.39       93.21
4                   5       91.79       97.19
5                   6       89.07       97.04
6                   7       94.72       91.07
7                   8       89.21       92.75

In [56]:
# Q:- Two catalysts are being analyzed to determine how they affect the mean yiled of a chemical process.
#Specifically,catalyst 1 is currently in use,but catalyst 2 is acceptable.
#Since catalyst 2 is cheaper, it should be adopted, providing it does not change the peocess yiled.
#A test is run in the pilot plNT nd results in the data shown in above table.
#Is there any difference B/W the mean yields?, Use 0.05, and assume equal variances.

In [57]:
#formulate the H0 & H1

**H0 :-** mu1 = mu2 [catalst 2 results in a mean yield that same from the mean yield when catalyst 1 is used]

**H1 :-** mu != mu2 [catalst 2 results in a mean yield that differs from the mean yield when catalyst 1 is used

In [58]:
A = list(df4.Catalyst_1)
A

[91.5, 94.18, 92.18, 95.39, 91.79, 89.07, 94.72, 89.21]

In [59]:
B = list(df4.Catalyst_2)
B

[89.19, 90.95, 90.46, 93.21, 97.19, 97.04, 91.07, 92.75]

In [60]:
stats.ttest_ind(A,B,equal_var =True)

Ttest_indResult(statistic=-0.3535908643461798, pvalue=0.7289136186068217)

In [61]:
#critival value

stats.t.ppf(0.025,14)  # [alpha/2, dof,(n1-1+n2-1)]

-2.1447866879169277

**Conclusion :**

       
   **Critival_Value_Approach:**
   
       -t_table_ value < t0 < +t_table_ value [-2.1447866879169277 < -0.3535908643461798 < 2.1447866879169277 ],so don not reject the NULL HYPOTHESIS.
       That is, at the 0.05 level of significance, we do not have strong evidence to conclude that catalyst 2, results in a mean yield that difference from the mean yield when catalyst 1 is used.
       
   **P-Value Approach:**
       
       
       p_value > alpha  so do not reject the null hypothesis,that is means re equal.
       

# Variace Unknown and assuned UNEQUAL

In [62]:
#NOTE :- The population variances ae assumed unequal, so a pooled variance isnot approproate.

#Use a t-value with v degreess of freedom, WHERE v =[(s1^2/n1)+(s2^2/n2)]/[[(s1^2)/n1)^2/(n1-1))+[(s2^2/n2)^2/(n2-1)]]

In [63]:
df5 = pd.read_excel('C:/Users/jagadeesh baman/Documents/alldata/sigma_unknown_assumed_unequal.xlsx')
df5

Metro_Phoenix  values_1 Rural_Arizona  values_2
0          Phoenix         3       Rimrock        48
1         Chandler         7      GoodYear        44
2          Gilbert        25     New_River        40
3             Mesa        10       Apachie        38
4  Paradise_valley        15       Buckeye        33
5           Peoria         6       Nogales        21
6      Scottsddale        12   Black_cayon        20
7            Tempe        25       Seadona        12
8        Sun _City        15        Payson         1
9             tnio         7   Casa_Grande        18

In [64]:
K = list(df5.values_1)
K

[3, 7, 25, 10, 15, 6, 12, 25, 15, 7]

In [65]:
J = list(df5.values_2)
J

[48, 44, 40, 38, 33, 21, 20, 12, 1, 18]

**P-value Approach**

In [66]:
stats.ttest_ind(K,J,equal_var=False)

Ttest_indResult(statistic=-2.7669395785560558, pvalue=0.015827284816100885)

**Critical Value Approach**

In [67]:
stats.t.ppf(0.025,13) #when you submite respected given values in the above given 'V', then you will get dof 13.2, so take 13

-2.160368656461013

**Conclusion**

     p-value < alpha & t-calculated is not B/W the t-table value, so REJECT THE H0
     
     
     
There is evidence to conclude that mean arsenic concentration in the frinking water in rural Arizona id=s different from the mean arsenic concentration in metropolitan Phoenix drinking water

# Dependent Samples

Test means of two related Populstions

        - Paired or matched samples
        -Repeated measures ( before/after)
        -Use difference between paired values:
        
           di - Xi -Yi
           
**Assumptions:**

        -Both populations are Normally Distributed.

In [68]:
# The test statistic for the mean difference is a t value,with n-1 degree of freedom:

              # t = (d(bar)-D0)/(sd/root(n))
    
#where D0 = hypothesized mean difference
#      sd = sample standard deviation of differences
#      n = the sample size (number of pairs)

In [69]:
#Q :- An article in the Journal of Strain Analysis compares sevral methods for predicting the shear strength for 
#steel plate girders
#Data for two of these methods, the Karlsruhe and lehigh procsdures,when applied to nine specific girders, are shown in Table.
#We wish to determine whether there is any difference (on the avereage) B/W the two methods.

In [70]:
df5=pd.read_excel('C:/Users/jagadeesh baman/Documents/alldata/strength_predictions_for_dependent_samples.xlsx')
df5

Girder   Karlsruhe_method  Lehigh_method
0     S11             1.186          1.061
1     S21             1.151          0.992
2     S31             1.322          1.063
3     S41             1.339          1.062
4     S51             1.200          1.065
5     S21             1.402          1.178
6     S22             1.365          1.037
7     S23             1.537          1.086
8     S24             1.559          1.052

In [71]:
#formulate the H0& H1

The parameter of interest is the difference in mean sher strength b/w the two methods,say, muD = mu1 - mu2 =0

   **HO :** mu1 = mu2
   
   **H1:** mu1 != mu2

In [72]:
KARL = list(df5.Karlsruhe_method)
KARL

[1.186, 1.151, 1.322, 1.339, 1.2, 1.402, 1.365, 1.537, 1.559]

In [73]:
LEH = list(df5.Lehigh_method)
LEH

[1.061, 0.992, 1.063, 1.062, 1.065, 1.178, 1.037, 1.086, 1.052]

In [74]:
stats.ttest_rel(KARL,LEH)

Ttest_relResult(statistic=6.0819394375848255, pvalue=0.00029529546278604066)

**Conclusion:** 
      
      
      Here P_value < alpha i.e [0.00029529546278604066 < 0.05 ], so we reject NULL HYPOTHESIS.
       
That at alpha 0.05  we can say that THERE IS DIFFERENCE(on the average ) B/W  the two methods.

# Inferences About the Difference B/W Two Population Proportions

* If the sample sizes are large, the samplinng distribution of p1(bar)-p2(bar) can be appproximated by a normal probability distribution.

* The sample aizes are sufficiently large if all of thsese conditions are met:

            n1p1 >= 5 ,n1(1-p1) >= 5
            
            
            n2p2 >= 5 ,n2(1-p2) >- 5

In [75]:
#Hypothesis

#we focus on tests involving no difference b/w the two population proportions ( i.e p1 =p2)

Q:- Extracts of St.John's Wort are widely used to trest depression.

* An article in the April issue of the Journal of the American Medical Association compared the efficacy of a standard extract of St.John's Wort with a Placedo in 200 outpatients diagnosed with major depression.

* Patients were randomly assigned to two groups; one goup received the St.John's Wort, and the other received the Placebo

* After 8 weeks,19 of the placebo-treated patients showed improvement, whereas 27 of those treated with St.John's Wort improved.

* Is there any reason to believe that St.John's Wort is effective in treating major depression, USE 0.05 Alpha

In [76]:
#Formulate The H0 & H1

**H0:** p1 = p2  [ i.e both are equal ]

**H1:** p1 != p2 [ i.e The claim that St.John's Wort is effective in treating major depression.]

In [77]:
#firstly define a function z_and p_value

In [78]:
import math
def two_samp_proportion(p1,p2,n1,n2):#it will take this as the input
    p_pool = ((p1*n2)+(p2*n1))/(n1+n2)
    x = (p_pool*(1-p_pool)*((1/n1)+(1/n2)))
    s=math.sqrt(x)
    z=(p1-p2)/s  #do no panic by looking this, this is just a formula, if you write in a paper it will become very easy.
    if(z < 0):
        p_val = stats.norm.cdf(z) #This is for left-tiled tesdt
    else:
        p_val =1-stats.norm.cdf(z)
    return z,p_val*2   #(*2 is because it is two-tail-test)

In [79]:
#Given
p1 = 0.27   #(27/100)
p2 = 0.19   #(19/100)
n1 = 100
n2 = 100

In [80]:
#call the defined function & give inputs
two_samp_proportion(0.27,0.19,100,100)

(1.3442056254198995, 0.17888190308175567)

In [81]:
stats.norm.cdf(1.3442056254198995)

0.9105590484591222

In [82]:
stats.norm.ppf(0.05/2)

-1.9599639845400545

**Conclusion:**

    Here z-calculated value falles into z_value regeion, so accept the null hypothesis.
    
**P_Value Approach:**

      Here P_value > Alpha [ 0.17888190308175567 > 0.05 ],Do not reject the NULL HYPOTHESIS.
      
 So There is insufficient evidence to support the claim that At.John's Wort is effective in treating major depression.

# Hypothesis Tests for Two Variances

In [83]:
#for this we use F-Test statistic  

#                        F=(s1^2/sigma1^2)/(s2^2/sigma2^2)
#Has an F distribution with (n1 - 1)numerator degrees of freedom and (n1 - 1) denominator degree of freedom

In [84]:
X3=[3,7,25,10,15,6,12,25,15,7]
Y3=[48,44,40,38,33,21,20,12,1,18]

In [85]:
Y3=[48,44,40,38,33,21,20,12,1,18]

In [86]:
F = np.var(X3)/np.var(Y3)
dfn=len(X3)-1
dfd=len(Y3)-1

In [87]:
p_value=scipy.stats.f.cdf(F,dfn,dfd)
p_value

0.024680183438910465

**Conclusion:**

        Reject H0 if p_value <= Alpha

# Z v/s T   Test

In [89]:
# 1:-  if n <= 30, and sigma Known then go for Z-Test, and if sigma known then go for t-test

# 2:- if n > 30, and sigma known then go for Z-test, if sigma unknown go for Z-test(here use sample standard deviation)

In [90]:
# Hope you found usfull,THANKYOU.